include/aos/common/thread.hpp

```diff
@@ -375,6 +386,158 @@ class ConditionalVariable : private NonCopyable {
     Error          mError;
 };
 
+/**
+ * Aos thread pool.
+ *
+ * @tparam cNumThreads number of threads used to perform tasks.
+ * @tparam cQueueSize tasks queue size.
+ * @tparam cMaxTaskSize max task size.
+ */
+template <size_t cNumThreads = 1, size_t cQueueSize = cDefaultThreadPoolQueueSize,
+    size_t cMaxTaskSize = cDefaultThreadPoolMaxTaskSize>
+class ThreadPool : private NonCopyable {
+public:
+    /**
+     * Creates thread pool instance.
+     */
+    ThreadPool()
+        : mCondVar(mMutex)
+        , mShutdown(false)
+        , mThreadsCount(0)
+    {
+    }
+
+    /**
+     * Adds task to task queue.
+     *
+     * @tparam T task type.
+     * @param functor task functor.
+     * @param arg argument passed to task functor when executed.
+     * @return Error.
+     */
+    template <typename T>
+    Error AddTask(T functor, void* arg = nullptr)
+    {
+        static_assert(sizeof(Function<T>) < cMaxTaskSize, "task is bigger than max task size");
+
+        LockGuard lock(mMutex);
+
+        auto err = mBuffer.PushValue(Function<T>(functor, arg));
+        if (!err.IsNone()) {
+            return err;
+        }
+
+        if (!(err = mCondVar.NotifyAll()).IsNone()) {
+            return err;
+        }
+
+        return ErrorEnum::eNone;
+    }
+
+    /**
+     * Runs thread pool.
+     *
+     * @return Error.
+     */
+    Error Run()
+    {
+        LockGuard lock(mMutex);
+
+        mShutdown = false;
+
+        for (auto& thread : mThreads) {
+            auto err = thread.Run([this](void*) {
+                uint8_t taskBuffer[cMaxTaskSize];
+
+                while (true) {
+                    UniqueLock lock(mMutex);
+
+                    auto err = mCondVar.Wait([this]() { return mShutdown || !mBuffer.IsEmpty(); });
+                    assert(err.IsNone());
+
+                    if (mShutdown) {
+                        return;
+                    }
+
+                    if (mBuffer.IsEmpty()) {
+                        continue;
+                    }
+
+                    err = mBuffer.Pop(taskBuffer, static_cast<CallableItf*>(mBuffer.Head())->Size());
+                    assert(err.IsNone());
+
+                    mThreadsCount++;
+
+                    lock.Unlock();
+
+                    (*static_cast<CallableItf*>(static_cast<void*>(taskBuffer)))();
+
+                    lock.Lock();
+
+                    mThreadsCount--;
+
+                    err = mCondVar.NotifyAll();
```

Comment:
Maybe we should use a separate condition variable for the Wait method. This way, we would only wake up the threads that are waiting for a specific condition, instead of waking up all the threads and making them check their condition. Otherwise, all the threads that are woken up will compete for the mutex.

Reply:
Indeed, it is good proposal. Will redo.